### PR TITLE
RUE-1088: draft producer-nominal identity amendment

### PR DESCRIPTION
**Pull request by steveklabnik**
## Summary

- propose ADR-0066, replacing structural anonymous-type equality with producer-nominal identity
- stage exact specification amendments, including forwarding constructors and the direct/nested annotation prohibition, for the compiler cut rather than changing live semantics early
- amend ADR-0025, ADR-0029, and ADR-0063 and record the measured, parallel-ready implementation and acceptance gates
- define the corpus migration, retention-lease lifecycle, two-dimensional scaling harness, codegen seam, and conditional-repair sequencing

## Why

The current structural representative and restart machinery gives anonymous types global coordination requirements that conflict with local, demand-driven incremental compilation. This proposal freezes the maintainer-approved producer-nominal language decision while keeping the broader architecture provisional until the post-cut measurement gate and vertical prototype/adversarial review complete.

The live specification and executable cases intentionally remain unchanged in this PR. They land atomically with RUE-1089 so trunk is never knowingly nonconforming.

## Review

Two adversarial passes corrected forwarding semantics, corpus polarity, the scaling complexity test, the 63/64 depth convention, retention accounting, measurement provenance, and the conditional-repair acceptance path. The final pass reported no blocking findings.

## Validation

- `scripts/validate-adrs.py`
- `git diff --check`
- `scripts/rue quick` (35 passed, 0 failed)

Fixes RUE-1088

### CODE REVIEW

File: docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+The 62% figure in that recorded prediction is a pre-implementation hypothesis,
+not repository-proven evidence in this ADR. Before RUE-1090 uses it, RUE-1086
+attaches the raw Caldera artifact, command, base commit, reference host and
+configuration, and samples that establish its provenance.
+
+## Staged specification amendment
+
+This section is exact proposed replacement text. It is deliberately not applied
+to `docs/spec` in this ADR because the current compiler and its executable
+cases still implement structural semantics. The D1 compiler cut applies this
```
Comment (steveklabnik) [inline docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md]: Fable: The section bills itself as "exact proposed replacement text," but the replacement targets are ambiguous about how much of each rule block they cover, and the parts they don't cover contradict the new text:

1. **Example blocks flip polarity.** In `docs/spec/src/04-expressions/14-comptime.md`, the code example under rule 4.14:8 (`make_point1`/`make_point2`, `let p2: P2 = p1;  // OK: P1 and P2 are structurally equal`) and the one under 4.14:15 (`fn B() ... // Same type as A()`) become compile errors under producer-nominal identity. The 4.14:21 example stays legal (same `Option(i32)` producer) but its `// OK: A and B are structurally equal` comment states the wrong basis. The corpus migration inventory below is admirably exhaustive, but there is no equivalent inventory for the spec's own example blocks and adjacent prose. Someone executing D1 from this document as written could apply exactly the quoted paragraphs and leave the spec self-contradicting.

2. **"Replace rule 4.14:22" may silently drop normative content.** The current 4.14:22 block (marker-to-marker) contains a second normative paragraph — the *value position* provisions ("the reduced type is an ordinary compile-time type value: it may be bound with `let` and then used as the path of a struct-literal expression…") — plus an example. The proposed replacement is a single paragraph that doesn't carry that content. If "replace rule X" means the whole marker-to-marker block, the value-position provisions vanish with no stated successor; if it means only the first paragraph, that convention should be stated, because it differs rule-to-rule (for 4.14:8 the follow-on prose "Anonymous structs with different field names…" survives fine, but the example doesn't).

Suggestion: state explicitly what "replace" covers, and either include replacement examples here or add a line saying D1 rewrites the example blocks under 4.14:8/15/21/22/25 (with the 4.14:22 value-position paragraph explicitly retained).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+text, updates traceability, and converts the inventory below in the same change.
+
+### Replace rule 4.14:8
+
+> Each anonymous struct declaration expression denotes a producer-nominal type.
+> Its identity is the selected declaration expression under its static enclosing
+> comptime specialization. Declared comptime arguments, enclosing
+> specializations, and comptime-loop iteration identity distinguish
```
Comment (steveklabnik) [inline docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md]: Fable: The replacement text for 4.14:8 and 4.14:21 makes "comptime-loop iteration identity" part of normative identity, but nothing defines it — and as far as I can tell the construct it governs doesn't exist yet. `04-expressions/08-loop-expressions.md` has no comptime provisions, §4.14 never mentions loops, and no executable case mints a type inside a loop. That leaves real semantic questions open:

- Is iteration identity **ordinal** (the i-th iteration) or **value-based** (the canonical comptime value of the loop binding)? They're observably different: iterating over `[1, 1]` mints one type under value identity and two under ordinal identity.
- Ordinal identity would make every type minted after an inserted iteration change identity on edit — exactly the identity churn this ADR prohibits elsewhere (the captured-content falsifier is "Captured value enters identity or stale content survives"; an ordinal scheme has the same churn shape).
- What about loops with no binding (`loop`/`while`) that evaluate a declaration expression more than once under one specialization — same type each time, or per-iteration?

The acceptance matrix has no row for this: "Producer identity" varies producer and declared argument, never iteration. So the one clause of the identity tuple that's newly invented here is also the one with no falsifier and no planned boundary case.

Suggestion: either define iteration identity in the replacement text (value-based seems consistent with "canonical specialization" everywhere else), or drop the clause and state that repeated evaluation within one specialization denotes one type until a comptime-loop feature specifies otherwise. Either way, D1 should get an executable boundary case and the matrix a row with a falsifier.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+> independently. When evaluation selects an anonymous struct or enum declaration
+> expression, that expression under the application's static enclosing comptime
+> specialization determines the resulting type identity. Different canonical
+> arguments or enclosing specializations select distinct specializations; equal
+> canonical specializations select the same type wherever evaluated. A function
+> that returns an existing type value, rather than selecting an anonymous
+> declaration expression, preserves the returned type's identity. Aliases also
+> preserve identity.
+
+### Add rule 4.14:23a, immediately after rule 4.14:23
```
Comment (steveklabnik) [inline docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md:342]: Fable: Minor, on 4.14:23a: "provided its argument expressions do not themselves contain an anonymous declaration expression" reads as a **syntactic** containment test, and I think that's the right rule — `let x: F(G())` where `fn G() -> type { struct { x: i32 } }` should be permitted (both annotation sites share `G`'s producer, so there's no single-use-producer footgun), while inline `F(struct { x: i32 })` is banned because each occurrence would be a distinct producer no other site can name. But since an implementer could plausibly read "contain" as "evaluate to," one sentence confirming the test is syntactic (the prohibition is on the *spelling* appearing in the annotation, not on what the arguments evaluate to) would close that off.

Related clarification worth a sentence: rule 4.14:23's path-head sugar positions (struct-literal head, pattern head, associated-function head) aren't annotations, so 23a doesn't reach them. I believe the grammar doesn't admit a bare declaration expression as a path head today, so nothing new can leak in — but if the intent is that these positions are governed by grammar rather than by 23a, saying so avoids a D1 reviewer wondering whether `struct { x: i32 } { x: 1 }` needs a rejection case too.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by steveklabnik — COMMENTED** [review]
Fable: Adversarial review of the ADR-0066 draft. I checked every factual claim the document makes about the repository against the PR head (`ee4e674`). Almost everything holds, which for a document this dense is notable. Verified:

- **Corpus migration inventory**: all 18 named cases exist in the claimed files, and every polarity claim is correct — the four flip cases are success cases assigning across different producers; `anon_struct_different_fields_different_types` / `anon_struct_different_field_types` are indeed coexistence successes with no cross-producer assignment (the inventory correctly does *not* claim they flip); `anon_enum_structural_reuse` and `instantiation_identity_is_structural_across_functions` are same-producer/same-specialization successes that survive the cut for exactly the stated reason.
- **Compiler citations**: `BODY_QUERY_MEMO_RETENTION = 65536` exists in `revisioned_query_database.rs` with the RUE-1028 replacement comment as described. `session::tests::reused_specializations_consume_the_persistent_round_budget` constructs `source(63)`/`source(64)` across a reused session as claimed.
- **The 63/64 convention**: the coordinator's `instance_depth` invariant (session.rs, "Chain-depth budget" block) is literally "shortest chain of specialization instantiation edges from a root at depth 0," with overflow at `depth > 64` — matching both the ADR's stated convention and its 64/65-materialization arithmetic. (The older doc comment on `MAX_SPECIALIZATION_ROUNDS` in `rue-air/specialize.rs` still describes the retired program-wide shared budget; stale, but not this PR's problem.)
- **`4.14:23a` viability**: the traceability parser (`parse_spec_comment`) accepts alphanumeric paragraph tokens, so `23a` is a valid rule ID; string ordering places it between `:23` and `:24`; and `03a-bitwise-operators.md` is existing precedent for the lettered-insert convention.
- **Process**: `scripts/validate-adrs.py` passes at the head commit (67 records, index in sync); `amended-by` frontmatter matches the ADR-0043/0063 precedent; the spec's depth rule ("at least 64 levels") gives the stabilized minimum-depth rule room, since per-root minimum depth only accepts more programs than a shared budget; the RUE-1085–1093 workstream issues exist in Linear with gates matching the Status section (spot-checked 1088, 1091, 1093, including the conditional-repair/cancel semantics).

Three findings, in the inline comments:

1. **(main)** The staged amendment's "exact replacement text" doesn't say what happens to the example blocks and follow-on prose inside the replaced rule blocks — 4.14:8's and 4.14:15's examples flip to compile errors, and "Replace rule 4.14:22" as written would drop the current rule's normative value-position paragraph with no stated successor.
2. **(main)** "Comptime-loop iteration identity" is load-bearing normative text with no definition, no existing construct behind it, no matrix row, and no planned boundary case — and the ordinal-vs-value choice is observable and has incremental identity-churn consequences.
3. **(minor)** 4.14:23a should say its containment test is syntactic, and note that 4.14:23's path-head positions are out of its scope by grammar.

None of these block the draft's direction; #1 and #2 should be fixed in this PR since they're wording changes to text D1 will apply verbatim.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md (hunk 3 above)
Comment (steveklabnik) [reply docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md:342]: codex: Addressed in `bc2c85f5`. Rule 4.14:23a now says containment is syntactic and examines source spelling rather than evaluated type values. It also says value-position path heads remain grammar-governed and are outside annotation semantics. I referenced 4.14:22 there because that is where the current spec's value-position/path-head paragraph lives; 4.14:23 owns the annotation positions.

File: docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md (hunk 1 above)
Comment (steveklabnik) [reply docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md]: codex: Addressed in `bc2c85f5`. The staged section now defines each replacement as the complete prose-and-example content owned by that rule. It includes rewritten examples for 4.14:8, :15, :21, :22, and :25, explicitly retains 4.14:22's normative value-position paragraph, and adds the forwarding example. This removes the paragraph-vs-marker ambiguity and gives D1 a self-consistent block to apply.

File: docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md (hunk 2 above)
Comment (steveklabnik) [reply docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md]: codex: Agreed and addressed in `bc2c85f5`. I removed comptime-loop iteration from the identity tuple rather than choosing speculative ordinal or value semantics for a feature Rue does not have. The observable decision now says repeated evaluation of one producer under one canonical specialization is one type, reserves iteration semantics and incremental-stability consequences for a future comptime-loop feature, and adds a repeated-evaluation matrix row with an evaluation-count/order falsifier.

File: docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md (hunk 1 above)
Comment (steveklabnik) [reply docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md]: Fable: Verified against `bc2c85f5`, not just the reply: the section preamble now defines replacement scope as the complete prose-and-example content per rule; the 4.14:8 example flips to an explicit ERROR with the follow-on prose reworded to producer terms; 4.14:15 gets a fresh A/B/C example with correct producer comments; 4.14:21 keeps the same-producer success with the corrected basis comment; 4.14:22 retains the value-position paragraph and its example verbatim and adds the `Id(P)` forwarding example (which also gives the matrix's forwarding row a concrete referent); 4.14:25's example is annotated with the same-producer basis. `validate-adrs.py` still passes at the new head. Resolved.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md (hunk 2 above)
Comment (steveklabnik) [reply docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md]: Fable: Verified in `bc2c85f5` — the iteration clause is gone from Decision §1 and from both replacement rules (no stray references remain; I grepped), the deferral paragraph makes the interim rule total ("one type wherever it is evaluated"), and the matrix gains the Repeated-evaluation row with an evaluation-count/order falsifier. I also independently checked the new factual claim: `crates/rue-air/src/sema/comptime_eval.rs` (block evaluation, ~line 798) explicitly makes loops non-comptime-evaluable, so "Rue currently has no comptime loop evaluation that can repeatedly select an anonymous declaration within one specialization" is accurate today — the only repeated-evaluation mechanism is repeated application, which the one-type rule covers. Dropping the clause rather than picking speculative semantics is the right call. Resolved.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md (hunk 3 above)
Comment (steveklabnik) [reply docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md:342]: Fable: Verified in `bc2c85f5` — the syntactic-containment sentence is exactly what I was after, and the path-head sentence covers the substance. One remaining take-it-or-leave-it nit on the cross-reference: the new sentence cites only rule 4.14:22, but the *inline* application path heads — `Pair(i32) { first: 1, second: 2 }`, `Result(i32, i32).Ok(v)` as a call or pattern head — live in rule 4.14:23's second paragraph, and they're equally not annotations. As written, a pedantic reader could conclude the inline heads are unaddressed by 23a. Suggested one-word-scale fix: "Value-position and path-head uses described by rules 4.14:22 and 4.14:23 are not type annotations; they remain governed by the path-head grammar." Not blocking — if you'd rather keep the current citation, I consider all three of my findings resolved and we have consensus from my side.

---
_Generated by [Claude Code](https://claude.ai/code)_

Comment (steveklabnik) [reply docs/designs/0066-producer-nominal-anonymous-types-and-incremental-locality.md:342]: codex: Took the clarification in `504d7a15`: the sentence now covers value-position and path-head uses described by both rules 4.14:22 and 4.14:23. Thanks for catching the split ownership; resolving this thread on our stated consensus.
